src/coreclr/jit/async.cpp

```diff
@@ -0,0 +1,1977 @@
+    LIR::AsRange(block).InsertAfter(callDefInfo.InsertAfter, continuationArg, storeContinuation);
+
+    GenTree* null                 = m_comp->gtNewNull();
+    GenTree* returnedContinuation = m_comp->gtNewLclvNode(m_returnedContinuationVar, TYP_REF);
+    GenTree* neNull               = m_comp->gtNewOperNode(GT_NE, TYP_INT, returnedContinuation, null);
+    GenTree* jtrue                = m_comp->gtNewOperNode(GT_JTRUE, TYP_VOID, neNull);
+
+    LIR::AsRange(block).InsertAfter(storeContinuation, null, returnedContinuation, neNull, jtrue);
+    *remainder = m_comp->fgSplitBlockAfterNode(block, jtrue);
+    JITDUMP("  Remainder is " FMT_BB "\n", (*remainder)->bbNum);
+
+    FlowEdge* retBBEdge = m_comp->fgAddRefPred(suspendBB, block);
+    block->SetCond(retBBEdge, block->GetTargetEdge());
+
+    block->GetTrueEdge()->setLikelihood(0);
+    block->GetFalseEdge()->setLikelihood(1);
+}
+
+//------------------------------------------------------------------------
+// AsyncTransformation::CreateResumption:
+//   Create the basic block that when branched to resumes execution on entry to
+//   the function.
+//
+// Parameters:
+//   block       - The block containing the async call
+//   remainder   - The block that contains the IR after the (split) async call
+//   call        - The async call
+//   callDefInfo - Information about the async call's definition
+//   stateNum    - State number assigned to this suspension point
+//   layout      - Layout information for the continuation object
+//
+// Returns:
+//   The new basic block that was created.
+//
+BasicBlock* AsyncTransformation::CreateResumption(BasicBlock*               block,
+                                                  BasicBlock*               remainder,
+                                                  GenTreeCall*              call,
+                                                  const CallDefinitionInfo& callDefInfo,
+                                                  unsigned                  stateNum,
+                                                  const ContinuationLayout& layout)
+{
+    if (m_lastResumptionBB == nullptr)
+    {
+        m_lastResumptionBB = m_comp->fgLastBBInMainFunction();
+    }
+
+    BasicBlock* resumeBB      = m_comp->fgNewBBafter(BBJ_ALWAYS, m_lastResumptionBB, true);
+    FlowEdge*   remainderEdge = m_comp->fgAddRefPred(remainder, resumeBB);
+
+    // It does not really make sense to inherit from the target, but given this
+    // is always 0% this just propagates the profile weight flag + sets
+    // BBF_RUN_RARELY.
+    resumeBB->inheritWeightPercentage(remainder, 0);
+    resumeBB->SetTargetEdge(remainderEdge);
+    resumeBB->clearTryIndex();
+    resumeBB->clearHndIndex();
+    resumeBB->SetFlags(BBF_ASYNC_RESUMPTION);
+    m_lastResumptionBB = resumeBB;
+
+    JITDUMP("  Creating resumption " FMT_BB " for state %u\n", resumeBB->bbNum, stateNum);
+
+    unsigned resumeByteArrLclNum = BAD_VAR_NUM;
+    if (layout.DataSize > 0)
```

Comment:
maybe leave a note that we need to restore non gc data first and gc data afterwards, so the gc data doesnt' get overwritten?